app/src/main.rs

```diff
@@ -64,8 +86,83 @@ async fn main() -> Result<()> {
         getresource_socket
     );
 
-    let keyprovider_server = grpc::keyprovider::start_service(keyprovider_socket);
-    let getresource_server = grpc::getresource::start_service(getresource_socket);
-
+    let keyprovider_server = rpc::keyprovider::start_grpc_service(keyprovider_socket);
+    let getresource_server = rpc::getresource::start_grpc_service(getresource_socket);
     tokio::join!(keyprovider_server, getresource_server).0
 }
+
+#[cfg(feature = "ttrpc")]
+fn main() {
+    env_logger::init();
+    let app_matches = App::new(rpc::AGENT_NAME)
+        .version(env!("CARGO_PKG_VERSION"))
+        .arg(
+            Arg::with_name("KeyProvider ttRPC Unix socket addr")
+                .long("keyprovider_sock")
+                .takes_value(true)
+                .help("This Unix socket address which the KeyProvider ttRPC service will listen to, for example: --keyprovider_sock unix:///tmp/aa_keyprovider",
+                ),
+        )
+        .arg(
+            Arg::with_name("GetResource ttRPC Unix socket addr")
+                .long("getresource_sock")
+                .takes_value(true)
+                .help("This Unix socket address which the GetResource ttRPC service will listen to, for example: --getresource_sock unix:///tmp/aa_getresource",
+                ),
+        )
+        .get_matches();
+
+    let keyprovider_socket = app_matches
+        .value_of("KeyProvider ttRPC Unix socket addr")
+        .unwrap_or(DEFAULT_KEYPROVIDER_UNIX_ADDR);
+
+    let getresource_socket = app_matches
+        .value_of("GetResource ttRPC Unix socket addr")
+        .unwrap_or(DEFAULT_GETRESOURCE_UNIX_ADDR);
+
+    debug!(
+        "KeyProvider ttRPC service listening on: {:?}",
+        keyprovider_socket
+    );
+    debug!(
+        "GetResource ttRPC service listening on: {:?}",
+        getresource_socket
+    );
+
+    if !Path::new(AA_UNIX_SOCKET_DIR).exists() {
+        std::fs::create_dir_all(AA_UNIX_SOCKET_DIR).expect("Creat AA socket directory failed");
+    }
+
+    clean_previous_sock_file(&keyprovider_socket).unwrap();
+    clean_previous_sock_file(&getresource_socket).unwrap();
+
+    let keyprovider_service = rpc::keyprovider::ttrpc_service();
+    let getresource_service = rpc::getresource::ttrpc_service();
+
+    let mut keyprovider_server = Server::new()
+        .bind(&keyprovider_socket)
+        .unwrap()
+        .register_service(keyprovider_service);
+    keyprovider_server.start().unwrap();
+
+    let mut getresource_server = Server::new()
+        .bind(&getresource_socket)
+        .unwrap()
+        .register_service(getresource_service);
+    getresource_server.start().unwrap();
+
+    loop {}
```

Comment:
Why we need a loop here?

Reply:
Here we need to hold the main process, or it will close the program directly.